File: php/testcode/fileupload_002.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_upload_uuid_rename
function fileupload002(BenchmarkRequest $req): BenchmarkResponse {
    $file = $req->file('f');
    if (!$file) {
        return BenchmarkResponse::badRequest('no file');
    }
    $ext = strtolower(pathinfo($file['name'], PATHINFO_EXTENSION));
    $allowed = ['jpg', 'png', 'gif'];
    if (!in_array($ext, $allowed, true)) { // vuln-code-snippet safe-line php_upload_uuid_rename
        return BenchmarkResponse::badRequest('invalid extension');
    }
    $dest = 'uploads/' . bin2hex(random_bytes(16)) . '.' . $ext;
    move_uploaded_file($file['tmp_name'], $dest);
    return BenchmarkResponse::ok("uploaded to $dest");
}
// vuln-code-snippet end php_upload_uuid_rename

File: ruby/testcode/fileupload_002.rb
require_relative 'shared'

ALLOWED_EXTENSIONS = %w[jpg png gif pdf].freeze

# vuln-code-snippet start ruby_fileupload_ext_allowlist
def upload_file_safe(req)
  upload = req.file('file')
  ext = File.extname(upload[:filename]).delete('.').downcase
  return BenchmarkResponse.bad_request('file type not allowed') unless ALLOWED_EXTENSIONS.include?(ext) # vuln-code-snippet safe-line ruby_fileupload_ext_allowlist
  File.write("/uploads/#{upload[:filename]}", upload[:data])
  BenchmarkResponse.ok('uploaded')
end
# vuln-code-snippet end ruby_fileupload_ext_allowlist
